src/lua/protobuf.lua

```diff
@@ -0,0 +1,662 @@
+    validate_table_is_array(field_def, value)
+    if is_scalar(field_def) then
+        local scalar_def = scalars[field_def.type]
+        encode_as_packed = scalar_def.encode_as_packed
+    end
+    for _, item in ipairs(value) do
+        local encoded_item = encode_field(protocol, field_def, item, true)
+        if encode_as_packed then
+            encoded_item = remove_tag(encoded_item)
+        end
+        table.insert(buf, encoded_item)
+    end
+    if encode_as_packed then
+        return wireformat.encode_len(field_def.id, table.concat(buf))
+    else
+        return table.concat(buf)
+    end
+end
+
+local function encode_enum(protocol, field_def, value)
+    local id = protocol[field_def.type]['id_by_value'][value]
+    if type(value) ~= 'number' and id == nil then
+        error(('%q is not defined in %q enum'):format(value, field_def.type))
+    end
+    -- According to open enums semantics unknown enum values are encoded as
+    -- numeric identifier. https://protobuf.dev/programming-guides/enum/
+    if type(value) == 'number' then
+        local subs_field_def = {type = 'int32', id = field_def.id}
+        validate_scalar(subs_field_def, value)
+        return scalars['int32'].encode(field_def.id, value)
+    else
+        return scalars['int32'].encode(field_def.id, id)
+    end
+end
+
+encode_field = function(protocol, field_def, value, ignore_repeated)
+    if field_def.repeated and not ignore_repeated then
+        return encode_repeated(protocol, field_def, value)
+    elseif is_scalar(field_def) then
+        validate_scalar(field_def, value)
+        local scalar_def = scalars[field_def.type]
+        return scalar_def.encode(field_def.id, value)
+    elseif is_enum(protocol, field_def) then
+        return encode_enum(protocol, field_def, value)
+    elseif is_message(protocol, field_def) then
+        local encoded_msg = encode(protocol, field_def.type, value)
+        validate_length(encoded_msg)
+        return wireformat.encode_len(field_def.id, encoded_msg, true)
+    else
+        assert(false)
+    end
+end
+
+-- Encodes the entered data in accordance with the
+-- selected protocol into binary format.
+--
+-- Accepts a protocol created by protobuf.protocol function,
+-- a name of a message selected for encoding and
+-- the data that needs to be encoded in the following format.
+--
+-- data = {
+--     <field_name> = <value>,
+--     <...>
+-- }
+encode = function(protocol, message_name, data)
+    local buf = {}
+    local message_def = protocol[message_name]
+    if message_def == nil then
+        error(('There is no message or enum named %q in the given protocol')
+            :format(message_name))
+    end
+    if message_def.type ~= 'message' then
+        assert(message_def.type == 'enum')
+        error(('Attempt to encode enum %q as a top level message'):format(
+            message_name))
+    end
+    local field_by_name = message_def.field_by_name
+    for field_name, value in pairs(data) do
```

Comment:
If a field is missing, this cycle skips it.

However, if a field has `box.NULL` value, it doesn't. What is a behavior in the case?

I think that it should be the same as for a missed field.

Reply:
Added a `box.NULL` handling same as missed field. Added test case. During testing mistakenly assigned `box.NULL` to already declared field and met 'Last One Wins' state: https://protobuf.dev/programming-guides/encoding/#last-one-wins. Will add test cases to check correctness of encoding in this state.